### ISSUE TITLE
Mapsui Offline Map

### ISSUE THREAD

**Issue by michalhoblik** [pauldendulk/Mapsui#248]
Hi pauldendulk,

first of all i would like to for your amazing job with Mapsui. I succesfully implemented your Mapsui in my project and everything is working fine. But my question is, and the next step in development of my App, is there possibility to load offline Map (is it mbtiles?) in Mapsui? I downloaded my mbtiles and succesfully loaded these mbtiles with "MbTilesTileSource" from Brutile and than created "VectorTileLayer" with this mbtile Source and added this Layer to the map. In Debug is everything fine. Mbtile has been loaded, layer created, datas are there but on the screen I see nothing. Just the white background screen and when I zoom it multiple times it crashes with unhandled exception. Am I doing it right ? Or if not can you give example, sample or anything how to load offline map in Mapsui? Here my code to create layer:
`        public static ILayer CreateTileLayer(string path)
        {
            var tileSource = new MbTilesTileSource(new SQLiteConnectionString(path, false), type: MbTilesType.BaseLayer);
            return new VectorTileLayer(tileSource) { Name = "MbTiles" };
        }`

Thank you for your time and answer.

Michal

**Comment by charlenni** [comment]
You are using a wrong MBTiles file. Mapsui could only handle bitmap MBTile format, not vector MBTile format. So try to download and use a bitmap MBTile file and try it again.

**Comment by michalhoblik** [comment]
Thank you charlenni,

and do you know or can you recommend me website or tool, where I can these bitmap MBTiles download or create? Everywhere I find just the vector mbtiles. I used mbtiles from this website Geofabrik: "http://download.geofabrik.de/europe/"

**Comment by charlenni** [comment]
You could get a demo at https://github.com/klokantech/tileserver-gl-data/blob/master/data/zurich-raster.mbtiles to look, if your code is correct. Than you could create raster MBTiles for example with MOBAC.

**Comment by michalhoblik** [comment]
Yes, that`s it. With zurich raster mbtiles example was everything all right. That means I need raster mbtiles. I downloaded MOBAC and I am going to play with it and hopefully I can create the map. Thanks for help.

**Comment by michalhoblik** [comment]
charlenni,
just one more question. Is there any option, that Mapsui will be in the future Vector mbtiles supporting ?

**Comment by charlenni** [comment]
That's a little bit difficult. If you want to have offline vector maps, you have to render them on the fly. For this, you need a reader, to read (in the best way) different vector formats. Than you need a styler, which delivers from a stylfile line color, width and so on for the tags, which are coming with the coordinates from the vector file. This both parts together could than be used to create a Mapsui feature, which than could be drawn on the screen :)

Do you create your app for a special platform?

**Comment by michalhoblik** [comment]
All right, thanks for the answer. Do you or paul planning something like this implement, because for the normal "not the GIS expert" seems this to be really heavy challange :). I just tried MOBAC you recommended me and so far I can tell I see just red crosses (not sure if it because of the proxy, i need to test it out at home :) ) and than I read that this tool has been banned on openstreetmap servers.

> Do you create your app for a special platform?

I'm developing my App with Xamarin Forms and for the platforms Android and iOS.

**Comment by 88madri** [comment]
You can use 
https://geopbs.codeplex.com/
to get MBTiles. (Remember that the use of MBTiles could be a copyright infringement)

**Comment by pauldendulk** [comment]
About vectortiles. There are some first steps but it will take a while before it is usable. The advantage of the first version will be smaller tiles and the option to change styling on the fly. For many scenarios this is not crucial and raster MBTiles will do.

I have moved this vectortile development to the vectortile branch now. I see you ran into it while it was in master. Master contains the latest release.

**Comment by pauldendulk** [comment]
About generating raster MBTiles. Mapsui has no tools to generate these. You should look for other sources as suggested by @charlenni and @88madri.

**Comment by pauldendulk** [comment]
What are you problems right now? There is a MBTiles sample in Mapsui.Samples.Wpf using some data in Spain. You got the Zurich sample to work. Now you are trying to get your own data into such a MBTiles?

**Comment by Sebastian1989101** [comment]
@michalhoblik Did you get it to work? I'm using Mapsui for a while now and currently have two Apps in the AppStore and two in the PlayStore which uses Xamarin.Forms and offline / local MBTiles file. So if you still have questions about it, I maybe able to help.

**Comment by pcloof** [comment]
You can also get a certain map extent offline programmatically. The basic approach is (in C#):

1. Obtain a HttpTileSource object (you can cast a TileLayer.TileSource to a HttpTileSource), suppose the object's name is 'X' for now.
2. Use X.Schema.GetTileInfos(...) to get a list of TileInfo for a given area and zoom level.
3. Call X.GetTile(...) on each TileInfo obtained in step 2. This gives you the raw bytes of each tile, provided that you have wifi or data enabled.
4. Store the necessary TileInfo properties and the raw tile data.
5. You can use the data later when offline. You create a TileProvider based on your tile data, put it in a TileSource and set this as source for your map layer.

Basically you use the same methods that are normally used under the bonnet while zooming and panning a map. Therefore, compatability of the data is self-evident and you don't need external tools. To use this correctly, you need to know the bounding box of the map area you need and the zoom level you want, so maybe you need some others websites or tools to determine these values in the language that Mapsui expects.

**Comment by Feroks** [comment]
@Sebastian1989101 Can you attach a screenshot of how the map looks in you applications?

**Comment by charlenni** [comment]
@Feroks I assume, he uses normal ratser MBTiles files for this. Than they look like all other raster tile files ;)

**Comment by pauldendulk** [comment]
@Feroks Here is one of Sebastian apps https://play.google.com/store/apps/details?id=com.SoftwareNotion.BreathCompanion&hl=en

There is a MBTilesSample in Mapsui.Samples.Wpf.

**Comment by pcloof** [comment]
I haven't been working on my app the last few weeks, so not really paying much attention to github, sorry for that. I use OpenStreetMap and basically get the tile layer with this code (step 1 of my previous post):

`using Mapsui.Layers;
using Mapsui.Utilities;
...
TileLayer tileLayer = OpenStreetMap.CreateTileLayer();
HttpTileSource tileSource = tileLayer.TileSource as HttpTileSource;`

**Comment by Visutronik** [comment]
@pauldendulk @pcloof 
Hello, 
i've tried to create my own tile but i have problems with steps 4 and 5.
I saved the byte array with a BinaryWriter as "mymap.mbtile" but i get an SQLite error "SQLite.SQLiteException: 'file is not a database'" if i try to open "mymap.mbtile". The "mymap.mbtile" file was created.

How can i save the byte array and TileInfo correctly?

The demo map from [https://github.com/klokantech/tileserver-gl-data/blob/master/data/zurich-raster.mbtiles](https://github.com/klokantech/tileserver-gl-data/blob/master/data/zurich-raster.mbtiles) works fine.

I try a german map from [http://osmlab.github.io/osm-qa-tiles/](http://osmlab.github.io/osm-qa-tiles/) but i see nothing.
